src/main/java/uk/nhs/hee/trainee/details/api/DspCredentialResource.java

```diff
@@ -0,0 +1,267 @@
+   * @param clientId                   the DSP client ID
+   * @param clientSecret               the DSP client secret
+   * @param redirectUri                the DSP redirect URI
+   * @param parEndpoint                the DSP Pushed Authorization Request (PAR) endpoint
+   * @param authorizeEndpoint          the DSP credential Authorize Request endpoint
+   * @param tokenEndpoint              the DSP Token Request endpoint
+   * @param restTemplateBuilder        the REST template builder
+   * @param jwtService                 the JWT service
+   */
+  public DspCredentialResource(PlacementService placementService,
+                               ProgrammeMembershipService programmeMembershipService,
+                               ObjectMapper objectMapper,
+                               @Value("${dsp.client-id}") String clientId,
+                               @Value("${dsp.client-secret}") String clientSecret,
+                               @Value("${dsp.redirect-uri}") String redirectUri,
+                               @Value("${dsp.par-endpoint}") String parEndpoint,
+                               @Value("${dsp.authorize-endpoint}") String authorizeEndpoint,
+                               @Value("${dsp.token.issue-endpoint}") String tokenEndpoint,
+                               RestTemplateBuilder restTemplateBuilder,
+                               JwtService jwtService) {
+    this.placementService = placementService;
+    this.programmeMembershipService = programmeMembershipService;
+    this.objectMapper = objectMapper;
+    this.clientId = clientId;
+    this.clientSecret = clientSecret;
+    this.redirectUri = redirectUri;
+    this.parEndpoint = parEndpoint;
+    this.issueEndpoint = authorizeEndpoint;
+    this.tokenEndpoint = tokenEndpoint;
+    this.restTemplateBuilder = restTemplateBuilder;
+    this.jwtService = jwtService;
+  }
+
+  /**
+   * Get the PAR response, including the request URI. We assume the trainee has been verified. This
+   * would be called by the front-end when the user clicks on the 'Add to wallet' button for a
+   * placement or programme membership.
+   *
+   * <p>NOTE: in future we will pass the content from the front-end as a JWT so that we can verify
+   * that it has not been tampered with.
+   *
+   * @param credentialType 'placement' or 'programmemembership'
+   * @param tisId The ID of the placement / programme membership.
+   * @return The PAR response, or Internal Server Error if the request times out
+   */
+  @GetMapping(value = "/par/{credentialType:placement|programmemembership}/{tisId}")
+  public ResponseEntity<ParResponse> getCredentialParUri(
+      @PathVariable(name = "credentialType") String credentialType,
+      @PathVariable(name = "tisId") String tisId,
+      @RequestHeader(HttpHeaders.AUTHORIZATION) String token) {
+    String traineeTisId;
+    log.info("Get credential request URI for {} with TIS ID {}", credentialType, tisId);
```

Comment:
We don't sanitise this to prevent log injection, but again this might be ok given the context (we don't sanitise in any of the other API calls either).